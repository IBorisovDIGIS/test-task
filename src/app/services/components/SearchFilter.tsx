import React, { ChangeEvent } from 'react';
import {
  Button, Grid, makeStyles, TextField,
} from '@material-ui/core';
import { useTranslation } from 'react-i18next';

const useStyles = makeStyles((theme) => ({
  margin: {
    margin: theme.spacing(1),
  },
  formControl: {
    flex: 1,
  },
  inputLabel: {
    position: 'relative',
    top: -25,
    padding: 0,
    margin: 0,
  },
  input: {
    borderRadius: 4,
    backgroundColor: theme.palette.common.white,
    border: '1px solid #ced4da',
    fontSize: 16,
    padding: '5px 10px',
  },
  button: {
    padding: '8px 10px',
  },
}));

interface ISearchFilter {
    handleChangeSearch: (value: string) => void;
    searchValue: string;
}

const SearchFilter: React.FC<ISearchFilter> = ({
  handleChangeSearch,
  searchValue,
}: ISearchFilter) => {
  const classes = useStyles();
  const { t } = useTranslation();

  const handleChangeFilter = (event: ChangeEvent<HTMLInputElement>) => {
    handleChangeSearch(event.target.value);
  };

  const handleResetFilter = () => handleChangeSearch('');

  return (
    <Grid container spacing={1}>
      <Grid item xs={4} container alignItems="stretch">
        <TextField
          fullWidth
          label={t('filter')}
          InputProps={{ disableUnderline: true, classes: { root: classes.input } }}
          InputLabelProps={{
            shrink: true,
          }}
          value={searchValue}
          onChange={handleChangeFilter}
        />
      </Grid>
      <Grid item xs={4} container alignItems="flex-end">
        <Button
          variant="outlined"
          classes={{ root: classes.button }}
          onClick={handleResetFilter}
        >
          {t('reset')}
        </Button>
      </Grid>
    </Grid>
  );
};

export default SearchFilter;
