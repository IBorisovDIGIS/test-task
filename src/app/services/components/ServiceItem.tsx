import React, { useState } from 'react';
import {
  Button, Grid, InputAdornment, makeStyles, Paper, TextField, Tooltip, Typography,
} from '@material-ui/core';
import copy from 'clipboard-copy';
import clsx from 'clsx';
import { useTranslation } from 'react-i18next';
import { black, blue, white } from '../../../constants/colors';
import CustomIcon from '../../common/CustomIcon';
import CopyIcon from '../../../assets/icons/copy.svg';
import { IBounce } from '../../../interfaces';

const useStyles = makeStyles((theme) => ({
  container: {
    padding: '26px 32px',
    marginTop: 24,
  },
  containerActivated: {
    backgroundColor: black,
  },
  button: {
    padding: '8px 10px',
  },
  title: {
    fontWeight: 300,
    fontSize: 29,
    color: '#000',
  },
  subtitle: {
    fontSize: 15,
    color: '#7D7D7D',
  },
  textActivated: {
    color: white,
  },
  bonusButton: {
    height: 48,
    backgroundColor: blue,
    color: white,
  },
  bonusButtonActivated: {
    backgroundColor: white,
    color: blue,
  },
  input: {
    borderRadius: 4,
    backgroundColor: theme.palette.common.white,
    border: '1px solid #ced4da',
    fontSize: 16,
    padding: '5px 10px',
  },
  copyIconWrapper: {
    height: '100%',
    width: 25,
  },
  copyIcon: {
    cursor: 'pointer',
  },
}));

const ServiceItem: React.FC<IBounce> = ({ description, promocode, title }:IBounce) => {
  const classes = useStyles();
  const { t } = useTranslation();

  const [isCopied, setIsCopied] = useState(false);
  const [isActivated, setIsActivated] = useState(false);

  const toggleBonus = () => setIsActivated((prevIsActivated) => !prevIsActivated);

  const handleOnCopy = async () => {
    await copy(promocode);
    setIsCopied(true);
    setTimeout(() => setIsCopied(false), 1000);
  };

  const renderCopyButton = () => (
    <Tooltip
      open={isCopied}
      title={promocode}
      placement="top"
      arrow
    >
      <InputAdornment
        position="end"
        onClick={handleOnCopy}
        className={classes.copyIconWrapper}
      >
        <CustomIcon
          icon={CopyIcon}
          iconClasses={classes.copyIcon}
        />
      </InputAdornment>
    </Tooltip>
  );

  return (
    <Grid item xs={12}>
      <Paper classes={{ root: clsx(classes.container, isActivated && classes.containerActivated) }}>
        <Grid container spacing={2} alignItems="stretch">
          <Grid item xs={6}>
            <Typography
              classes={{ root: clsx(classes.title, isActivated && classes.textActivated) }}
            >
              {title}
            </Typography>
            <Typography
              classes={{ root: clsx(classes.subtitle, isActivated && classes.textActivated) }}
            >
              {description}
            </Typography>
          </Grid>
          <Grid item xs={3} container alignItems="flex-end">
            <TextField
              fullWidth
              label={t('promocode').toUpperCase()}
              disabled
              value={promocode}
              InputProps={{
                disableUnderline: true,
                classes: { root: classes.input },
                endAdornment: renderCopyButton(),
              }}
            />
          </Grid>
          <Grid item xs={3} container alignItems="flex-end">
            <Button
              variant="contained"
              fullWidth
              classes={{
                root: clsx(classes.bonusButton, isActivated && classes.bonusButtonActivated),
              }}
              onClick={toggleBonus}
            >
              {isActivated ? t('deActivateBonus') : t('activateBonus')}
            </Button>
          </Grid>
        </Grid>
      </Paper>
    </Grid>
  );
};

export default ServiceItem;
