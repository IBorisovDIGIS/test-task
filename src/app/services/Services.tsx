import React, { useCallback, useEffect } from 'react';
import { Grid } from '@material-ui/core';
import SearchFilter from './components/SearchFilter';
import ServiceItem from './components/ServiceItem';
import { IBounce } from '../../interfaces';

interface IServices {
    bonuses: IBounce[]
}

const Services: React.FC<IServices> = ({ bonuses }: IServices) => {
  const [searchValue, setSearchValue] = React.useState('');
  const [filteredBounces, setFilteredBounces] = React.useState(bonuses);

  const handleFilterBounces = useCallback(() => {
    setFilteredBounces(() => bonuses.filter(
      ({ title }) => title.toLowerCase().includes(searchValue.toLowerCase()),
    ));
  }, [setFilteredBounces, searchValue]);

  useEffect(() => {
    handleFilterBounces();
  }, [searchValue, bonuses]);

  const handleChangeSearch = useCallback((value: string) => {
    setSearchValue(value);
  }, [searchValue]);

  return (
    <Grid container>
      <Grid item xs={12}>
        <SearchFilter
          handleChangeSearch={handleChangeSearch}
          searchValue={searchValue}
        />
      </Grid>
      <Grid container item xs={12}>
        {filteredBounces.map((bounce) => <ServiceItem {...bounce} key={bounce.title} />)}
      </Grid>
    </Grid>
  );
};

export default Services;
