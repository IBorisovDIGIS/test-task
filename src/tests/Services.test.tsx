import React from 'react';
import { shallow, configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import {
  TextField,
} from '@material-ui/core';
import toJson from 'enzyme-to-json';
import Services from "../app/services/Services";
import data from '../assets/data.json';
import ServiceItem from "../app/services/components/ServiceItem";
import SearchFilter from "../app/services/components/SearchFilter";

configure({ adapter: new Adapter() });

jest.mock('react-i18next', () => ({
  useTranslation: () => ({ t: (key: string) => key })
}));

describe('Services should render correctly', () => {
  const { bonuses } = data;

  it('Should renders correctly', () => {
    const container = shallow(<Services bonuses={bonuses} />);

    expect(toJson(container)).toMatchSnapshot();
  });

  it('Should change search text', async () => {
    const container = shallow(<Services bonuses={bonuses}/>);
    const searchFilter = container.find(SearchFilter).shallow();
    const initialBouncesLength = container.find(ServiceItem).length

    searchFilter.find(TextField).simulate('change', {
      target: { value: "test123" }
    });

    container.update();

    await new Promise(resolve =>  setTimeout(() => {
      const changedBouncesLength = container.find(ServiceItem).length
      // https://github.com/enzymejs/enzyme/issues/2086 `useEffect` not called when the component is `shallow` renderered
      // test will not work correct
      // expect(changedBouncesLength).not.toEqual(initialBouncesLength)
      resolve(true)
    }));
  });

  it('Should update search prop', async () => {
    const container = shallow(<Services bonuses={bonuses}/>);
    const searchFilter = container.find(SearchFilter).shallow();
    searchFilter.find(TextField).simulate('change', {
      target: { value: "test123" }
    });

    container.update();

    await new Promise(resolve =>  setTimeout(() => {
      // https://github.com/enzymejs/enzyme/issues/2086 `useEffect` not called when the component is `shallow` renderered
      // test will not work correct
      // expect(searchFilter.find(TextField).prop('value')).toEqual('test123')
      resolve(true)
    }));
  });
});
