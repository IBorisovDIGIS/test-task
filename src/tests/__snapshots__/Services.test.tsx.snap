// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Services should render correctly Should renders correctly 1`] = `
<WithStyles(ForwardRef(Grid))
  container={true}
>
  <WithStyles(ForwardRef(Grid))
    item={true}
    xs={12}
  >
    <SearchFilter
      handleChangeSearch={[Function]}
      searchValue=""
    />
  </WithStyles(ForwardRef(Grid))>
  <WithStyles(ForwardRef(Grid))
    container={true}
    item={true}
    xs={12}
  >
    <ServiceItem
      description="Site constructor"
      key="Sitecostructor.io"
      link="https://sitecostructor.io/?p=itpaycodes"
      promocode="itpaycodes"
      title="Sitecostructor.io"
    />
    <ServiceItem
      description="SDK"
      key="Appvision.com"
      link="http://appvision.com/+itpaycodes"
      promocode="itpaycodes"
      title="Appvision.com"
    />
    <ServiceItem
      description="Analytics"
      key="Analytics.com"
      link="https://www.analytics.com/?ref=itpaycodes"
      promocode="itpaycodes"
      title="Analytics.com"
    />
    <ServiceItem
      description="Logos"
      key="Logotype"
      link="https://www.logotype.design/"
      promocode="itpaycodes"
      title="Logotype"
    />
    <ServiceItem
      description="SSL"
      key="Certificates"
      link="https://www.certificates.com/ssl-certificate/"
      promocode="itpaycodes"
      title="Certificates"
    />
    <ServiceItem
      description="SEO"
      key="Seorank.com"
      link="https://seorank.com/"
      promocode="itpaycodes"
      title="Seorank.com"
    />
    <ServiceItem
      description="Marketing"
      key="marke.com"
      link="http://marketing.com/"
      promocode="itpaycodes"
      title="marke.com"
    />
    <ServiceItem
      description="Hosting"
      key="Major.com"
      link="https://www.hosting.com/"
      promocode="itpaycodes"
      title="Major.com"
    />
    <ServiceItem
      description="Web sites"
      key="Setup.com"
      link="http://www.setup.com/instructions/promocode"
      promocode="itpaycodes"
      title="Setup.com"
    />
    <ServiceItem
      description="Analytics"
      key="topvisor.com"
      link="https://topvisor.com/promo/itpaycodes/"
      promocode="itpaycodes"
      title="topvisor.com"
    />
    <ServiceItem
      description="SEO"
      key="seopult.com"
      link="https://promopult.com/"
      promocode="itpaycodes"
      title="seopult.com"
    />
    <ServiceItem
      description="Emails"
      key="sendpulse.com"
      link="https://sendpulse.com/"
      promocode="itpaycodes"
      title="sendpulse.com"
    />
    <ServiceItem
      description="Content marketing"
      key="Turbotext.com"
      link="http://turbotext.com/"
      promocode="itpaycodes"
      title="Turbotext.com"
    />
    <ServiceItem
      description="Widgets"
      key="Witget.com"
      link="https://witget.com/?partner=itpaycodes"
      promocode="itpaycodes"
      title="Witget.com"
    />
    <ServiceItem
      description="Analytics"
      key="Spywords.com"
      link="http://spywords.com"
      promocode="itpaycodes"
      title="Spywords.com"
    />
    <ServiceItem
      description="Content Marketing"
      key="Etxt.com"
      link="https://www.etxt.com/"
      promocode="itpaycodes"
      title="Etxt.com"
    />
    <ServiceItem
      description="Analytics"
      key="Serpstat.com"
      link="https://analytics.com"
      promocode="itpaycodes"
      title="Serpstat.com"
    />
    <ServiceItem
      description="Hosting"
      key="webhost.com"
      link="https://webhost.com/"
      promocode="itpaycodes"
      title="webhost.com"
    />
    <ServiceItem
      description="Email marketing"
      key="Unisender.com"
      link="https://www.unisender.com"
      promocode="itpaycodes"
      title="Unisender.com"
    />
    <ServiceItem
      description="Advertisment"
      key="getuniq.com"
      link="http://itpaycodes.getuniq.com"
      promocode="itpaycodes"
      title="getuniq.com"
    />
  </WithStyles(ForwardRef(Grid))>
</WithStyles(ForwardRef(Grid))>
`;
